src/Disks/DiskLocal.cpp

```diff
@@ -303,14 +312,26 @@ bool inline isSameDiskType(const IDisk & one, const IDisk & another)
 void DiskLocal::copy(const String & from_path, const std::shared_ptr<IDisk> & to_disk, const String & to_path)
 {
     if (isSameDiskType(*this, *to_disk))
-        Poco::File(disk_path + from_path).copyTo(to_disk->getPath() + to_path); /// Use more optimal way.
+    {
+        fs::path to = fs::path(to_disk->getPath()) / to_path;
+        fs::path from;
+        if (from_path.ends_with('/'))
+            from = fs::path(disk_path) / from_path.substr(0, from_path.size() - 1);
+        else
+            from = fs::path(disk_path) / from_path;
+
+        if (fs::is_directory(from))
+            to /= from.filename();
```

Comment:
> may return empty string if to_path.ends_with('/')

probably you ment here not `to_path`, but `from_path`? (because `from.filename()` ends with `from_path`)
`from.filename` can be an empty string here only of `from_path == "/"` (because there is a check for that with modification to `parent_path` above). Then it will be ok.